### PR TITLE
Deepen SearchConfig with a solver-timeout resolution seam

### PR DESCRIPTION
**Pull request by pmatos**
## Refactoring goal

Architecture audit (via the `improve-codebase-architecture` skill) surfaced a **shallow, duplicated, and partially-buggy** surface: every SMT-driven search backend re-derived the per-query Z3 solver timeout, and the "clamp it to the remaining search budget" rule was implemented in only half of them.

This deepens that policy into a single tested seam on `SearchConfig`.

### The friction (before)

The rule *"resolve the per-query solver timeout (fall back to 5s when unset), then clamp it to the time left in the overall search budget"* lived in four places:

| Backend | resolves 5s fallback | clamps to remaining budget |
|---|---|---|
| `enumerative` | ✅ | ✅ |
| `llm` | ✅ | ✅ |
| `stochastic` (MCMC) | ✅ | ❌ |
| `symbolic` | ✅ | ❌ |

Consequences:
- **Latent correctness bug.** `stochastic` computed the solver timeout *once* before the loop and reused it every iteration; `symbolic` resolved it per query. Neither clamped to the remaining budget, so a nearly-exhausted `--timeout` could still hand Z3 the full solver timeout (default **30s**) on its final query and overrun the deadline.
- **Duplication.** The `5s` fallback literal was spelled out in all four backends, while the config default ships `30s` — one concept, several spellings.

### The deepening (after)

Two methods on `SearchConfig` (`src/search/config.rs`):

- `solver_timeout() -> Duration` — the fallback resolution; the single home for the `5s` literal.
- `solver_timeout_within_budget(elapsed) -> Option<Duration>` — resolution **+** clamp to the remaining budget **+** sub-millisecond exhaustion (returns `None` when no Z3-representable budget remains).

All four backends now resolve their per-query timeout through this seam:
- `stochastic` and `symbolic` **inherit the budget clamp** (the bug fix).
- `enumerative` and `llm` are **behaviour-preserving** — their sub-ms and 1ms floors already coincided, so routing them through the shared rule changes nothing observable.

**Deletion test:** delete the (formerly enumerative-only) helper and the clamp logic reappears — and was *missing* — across the other backends. It earns its keep at N=4.

## Tests (TDD, red → green)

Written test-first against the new seam's interface, then implemented to pass:

- `solver_timeout_resolves_default_fallback` — default (`30s`), explicit, and `None → 5s` fallback.
- `solver_timeout_within_budget_clamps_to_remaining_search_time` — solver-below-budget passthrough, clamp-to-remaining, exact exhaustion, sub-millisecond exhaustion, over-elapsed.
- `solver_timeout_within_budget_is_unbounded_without_search_timeout` — no overall timeout ⇒ solver timeout returned unclamped.
- `symbolic_smt_timeout_is_clamped_to_remaining_search_budget` — **behavioural proof of the fix**: with a 30s solver timeout and a 50ms search budget, the timeout handed to Z3 (recorded via the `TestIsa` mock) is clamped to ~50ms, not 30s.

The prior enumerative unit test that pinned this logic was migrated onto the new seam (its coverage is fully preserved by the `config.rs` tests).

## Verification

- `./ci_check.sh` passes end to end: `cargo fmt --check`, build, AArch64 test-binary build, full `cargo test` (unit + integration), and `test_all.sh`.
- `cargo clippy --all-targets` introduces no findings in the touched files.
- Full lib suite: 1457 passing.

No public CLI behaviour changes; the only observable change is that stochastic and symbolic search now respect the search deadline when timing out Z3.

### CODE REVIEW

File: src/search/config.rs [hunk 1]
```diff
@@ -487,6 +487,40 @@ impl SearchConfig {
...
+    ///
+    /// All SMT-driven backends resolve their per-query timeout here so the
+    /// budget-clamping rule cannot drift between them.
+    pub fn solver_timeout_within_budget(&self, elapsed: Duration) -> Option<Duration> {
+        let solver_timeout = self.solver_timeout();
+        let timeout = match self.timeout {
+            Some(search_timeout) => solver_timeout.min(search_timeout.checked_sub(elapsed)?),
+            None => solver_timeout,
+        };
+        (timeout.as_millis() > 0).then_some(timeout)
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/search/config.rs:522]: **Out of scope for this PR** — this is a real, reachable behaviour change, but "0 = unbounded" was never a coherent cross-backend contract: pre-#669 only symbolic and stochastic passed `Duration::ZERO` through to Z3, while `enumerative` already used the same `as_millis() > 0` gate and `llm` the `>= MIN_SMT_TIMEOUT` gate — both already treated `0` as "skip". Restoring "0 = unbounded" would also fight this PR's whole purpose: an unbounded Z3 query can overrun the overall `--timeout` deadline, the exact bug this PR fixes. Distinguishing a *configured* zero (unbounded) from a *clamped-to-zero* remaining budget needs a deliberate policy decision (reject `0` at the CLI, add an explicit unbounded sentinel, or document skip) and a new seam semantic, so it belongs in a follow-up rather than this seam-unification change.

Two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.6) both classified this as a deferral. Tracked as #670 (https://github.com/pmatos/s11/issues/670), which also notes that this PR's "no public CLI behaviour changes" claim should be qualified for the `--solver-timeout 0` edge case.

**Comment by pmatos** [comment]
Regarding the review's one concrete suggestion (the optional MCMC-level behavioral test):

**Fixed in `8fbf8fe`.**

Changed: added `stochastic_smt_timeout_is_clamped_to_remaining_search_budget` to `src/search/stochastic/mcmc.rs`. It drives `MCMCSearch::search` with a tight 50ms overall `--timeout` and a 30s `--solver-timeout` through the deterministic `TimeoutProbeIsa` probe, then asserts the timeout handed to the solver is clamped to the ~50ms remaining budget (`1..=50`ms) rather than the full 30s. This mirrors `symbolic_smt_timeout_is_clamped_to_remaining_search_budget` and closes the MCMC-level coverage gap you flagged — the stochastic half of the deadline fix was previously exercised only transitively through the shared `SearchConfig` seam unit tests, never via `MCMCSearch::search` itself. Per your guidance it uses the existing deterministic probe, not fragile wall-clock timing.

Validation: `cargo fmt -- --check` (clean), full lib suite (**1458 passing**, including the new test), `cargo clippy --all-targets` (no new warnings from the added code), and a `RUSTFLAGS="-D warnings"` build all pass locally.

**Comment by pmatos** [comment]
Thanks for the re-review — glad the MCMC behavioral test reads as a solid proof of the fix. The three observations you raised are all explicitly non-blocking, and after evaluation none warrants a code change or a tracking issue:

**1. mcmc `break` placement — no change.** The new `break` correctly guards only the SMT-query path (inside `if proposal_cost < best_cost`), so on an exhausted budget with *no* cheaper proposal no query is issued anyway and the top-of-loop `config.timeout` check — which runs before any work every iteration — exits on the next pass. The two checks measure different things (per-query SMT budget vs. overall wall-clock deadline), so the sub-millisecond "disagreement" is intentional cooperative-check granularity, not a defect. Matches your own "not worth restructuring."

**2. llm module doc — no change.** The intent is already self-evident at the call site: `verification_timeout_for_remaining` visibly keeps its own `remaining.min(solver_timeout)` clamp and only adopts `config.solver_timeout()` for the shared 5s fallback, and the PR description already states llm is behaviour-preserving. There's no project convention requiring a module-doc note for partial seam adoption.

**3. Test call-site formatting — no change.** `cargo fmt -- --check` passes on the tree; rustfmt is the authority on wrapping, so the multi-line call sites are correct by definition.

_This assessment was made by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.6). If you disagree, please reply and we'll re-evaluate._
